File: crates/shell/src/cmds/env.rs
// SPDX-License-Identifier: GPL-2.0-only
//
// Keira Kernel - Operating System Kernel
// Copyright (C) 2026 Moh. Ananda Firmansyah Putra
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.

#![allow(unused_variables, unused_unsafe)]

//!
//! View and modify system environment variables.

use crate::args::CliArgs;
use crate::state::*;
use keira_io::vga;

pub fn run(parts: &mut core::str::SplitWhitespace) {
    let args = CliArgs::parse(parts);

    if args.has_flag('h', "help") {
        unsafe {
            vga::set_color(vga::Color::White, vga::Color::Black);
            vga::print_str("Usage: env [key] [value] [-l] [-u <key>]\n\n");
            vga::print_str("Description:\n  View or modify environment variables ($USER, $HOME, $PATH, $SHELL).\n\n");
            vga::print_str("Options:\n");
            vga::print_str("  -l, --list     List all active environment variables\n");
            vga::print_str("  -u, --unset    Unset the specified environment variable\n");
            vga::print_str("  -h, --help     Show this help message and exit\n");
            vga::set_color(vga::Color::LightGrey, vga::Color::Black);
        }
        return;
    }

    unsafe {
        if let Some(unset_key) = args.get_opt('u', "unset") {
            let _ = set_env_var(unset_key, "");
            vga::set_color(vga::Color::LightGreen, vga::Color::Black);
            vga::print_str("[OK] Unset environment variable: ");
            vga::print_str(unset_key);
            vga::print_str("\n");
            vga::set_color(vga::Color::LightGrey, vga::Color::Black);
            return;
        }

        let key = args.first_positional();
        let val = args.second_positional();

        match (key, val) {
            (Some(k), Some(v)) => {
                if set_env_var(k, v).is_ok() {
                    vga::set_color(vga::Color::LightGreen, vga::Color::Black);
                    vga::print_str("[OK] Updated environment: ");
                    vga::print_str(k);
                    vga::print_str("=");
                    vga::print_str(v);
                    vga::print_str("\n");
                } else {
                    vga::set_color(vga::Color::LightRed, vga::Color::Black);
                    vga::print_str("Error: Failed to set environment variable.\n");
                }
                vga::set_color(vga::Color::LightGrey, vga::Color::Black);
            }
            (Some(k), None) => {
                if let Some(eq_pos) = k.find('=') {
                    let key_part = &k[..eq_pos];
                    let val_part = &k[eq_pos + 1..];
                    if set_env_var(key_part, val_part).is_ok() {
                        vga::set_color(vga::Color::LightGreen, vga::Color::Black);
                        vga::print_str("[OK] Set environment: ");
                        vga::print_str(key_part);
                        vga::print_str("=");
                        vga::print_str(val_part);
                        vga::print_str("\n");
                    } else {
                        vga::set_color(vga::Color::LightRed, vga::Color::Black);
                        vga::print_str("Error: Failed to set environment variable.\n");
                    }
                    vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                    return;
                }

                let mut buf = [0u8; 128];
                if let Ok(len) = get_env_var(k, &mut buf) {
                    vga::set_color(vga::Color::White, vga::Color::Black);
                    vga::print_str(k);
                    vga::print_str("=");
                    vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                    if let Ok(s) = core::str::from_utf8(&buf[..len]) {
                        vga::print_str(s);
                    }
                    vga::print_str("\n");
                } else {
                    vga::set_color(vga::Color::LightRed, vga::Color::Black);
                    vga::print_str("Environment variable not found: ");
                    vga::print_str(k);
                    vga::print_str("\n");
                }
                vga::set_color(vga::Color::LightGrey, vga::Color::Black);
            }
            _ => {
                vga::set_color(vga::Color::White, vga::Color::Black);
                vga::print_str("Active Environment Variables:\n");
                vga::set_color(vga::Color::LightGrey, vga::Color::Black);

                let default_keys = [
                    "USER", "HOME", "PATH", "SHELL", "TERM", "LANG", "KERNEL", "OSTYPE",
                    "HOSTTYPE",
                ];
                for k in default_keys.iter() {
                    let mut buf = [0u8; 128];
                    if let Ok(len) = get_env_var(k, &mut buf) {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  ");
                        vga::print_str(k);
                        vga::print_str("=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        if let Ok(s) = core::str::from_utf8(&buf[..len]) {
                            vga::print_str(s);
                        }
                        vga::print_str("\n");
                    } else if *k == "TERM" {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  TERM=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        vga::print_str("xterm-256color\n");
                    } else if *k == "LANG" {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  LANG=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        vga::print_str("en_US.UTF-8\n");
                    } else if *k == "KERNEL" {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  KERNEL=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        vga::print_str("Keira\n");
                    } else if *k == "OSTYPE" {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  OSTYPE=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        vga::print_str("keira-kernel\n");
                    } else if *k == "HOSTTYPE" {
                        vga::set_color(vga::Color::White, vga::Color::Black);
                        vga::print_str("  HOSTTYPE=");
                        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
                        vga::print_str("x86_64\n");
                    }
                }

                vga::set_color(vga::Color::LightGrey, vga::Color::Black);
            }
        }
    }
}

File: userland/lib/stdlib/env/env.c
/* SPDX-License-Identifier: GPL-2.0-only */
/*
 * Keira Kernel - Freestanding Kernel from Scratch
 * Copyright (C) 2026 Moh. Ananda Firmansyah Putra
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 */

#include <errno.h>
#include <malloc.h>
#include <stdlib.h>
#include <string.h>

char **environ = NULL;

char *getenv(const char *name) {
    if (!name || !environ)
        return NULL;
    size_t len = strlen(name);
    for (char **env = environ; *env != NULL; env++) {
        if (strncmp(*env, name, len) == 0 && (*env)[len] == '=') {
            return *env + len + 1;
        }
    }
    return NULL;
}

int setenv(const char *name, const char *value, int overwrite) {
    if (!name || *name == '\0' || strchr(name, '=') != NULL) {
        errno = EINVAL;
        return -1;
    }
    if (!value)
        value = "";

    char *existing = getenv(name);
    if (existing && !overwrite)
        return 0;

    size_t name_len = strlen(name);
    size_t val_len = strlen(value);
    char *new_entry = (char *)malloc(name_len + 1 + val_len + 1);
    if (!new_entry) {
        errno = ENOMEM;
        return -1;
    }
    memcpy(new_entry, name, name_len);
    new_entry[name_len] = '=';
    memcpy(new_entry + name_len + 1, value, val_len);
    new_entry[name_len + 1 + val_len] = '\0';

    if (existing) {
        for (char **env = environ; *env != NULL; env++) {
            if (strncmp(*env, name, name_len) == 0 && (*env)[name_len] == '=') {
                *env = new_entry;
                return 0;
            }
        }
    }

    size_t count = 0;
    if (environ) {
        while (environ[count] != NULL)
            count++;
    }

    char **new_environ = (char **)malloc((count + 2) * sizeof(char *));
    if (!new_environ) {
        free(new_entry);
        errno = ENOMEM;
        return -1;
    }
    for (size_t i = 0; i < count; i++) {
        new_environ[i] = environ[i];
    }
    new_environ[count] = new_entry;
    new_environ[count + 1] = NULL;
    environ = new_environ;
    return 0;
}

int unsetenv(const char *name) {
    if (!name || *name == '\0' || strchr(name, '=') != NULL) {
        errno = EINVAL;
        return -1;
    }
    if (!environ)
        return 0;

    size_t len = strlen(name);
    char **src = environ;
    char **dst = environ;
    while (*src) {
        if (strncmp(*src, name, len) == 0 && (*src)[len] == '=') {
            src++;
        } else {
            *dst++ = *src++;
        }
    }
    *dst = NULL;
    return 0;
}

int putenv(char *string) {
    if (!string || strchr(string, '=') == NULL) {
        errno = EINVAL;
        return -1;
    }
    char *eq = strchr(string, '=');
    size_t name_len = (size_t)(eq - string);
    for (char **env = environ; env && *env != NULL; env++) {
        if (strncmp(*env, string, name_len) == 0 && (*env)[name_len] == '=') {
            *env = string;
            return 0;
        }
    }

    size_t count = 0;
    if (environ) {
        while (environ[count] != NULL)
            count++;
    }
    char **new_environ = (char **)malloc((count + 2) * sizeof(char *));
    if (!new_environ) {
        errno = ENOMEM;
        return -1;
    }
    for (size_t i = 0; i < count; i++) {
        new_environ[i] = environ[i];
    }
    new_environ[count] = string;
    new_environ[count + 1] = NULL;
    environ = new_environ;
    return 0;
}
